server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -307,7 +308,9 @@ public void markAsDone() {
             try {
                 // this might still throw an exception ie. if the shard is CLOSED due to some other event.
                 // it's safer to decrement the reference in a try finally here.
-                indexShard.postRecovery("peer recovery done");
+                PlainActionFuture<Void> future = PlainActionFuture.newFuture();
+                indexShard.postRecovery("peer recovery done", future);
+                future.actionGet();
```

Comment:
`markAsDone()` is kind of a fire-and-forget thing (I'm not even seeing anywhere that handles the exception it might throw) so I think we can make this nonblocking fairly easily too:

```java
    public void markAsDone() {
        if (finished.compareAndSet(false, true)) {
            assert multiFileWriter.tempFileNames.isEmpty() : "not all temporary files are renamed";
            indexShard.postRecovery("peer recovery done", ActionListener.runBefore(new ActionListener<>() {
                @Override
                public void onResponse(Void unused) {
                    listener.onRecoveryDone(state(), indexShard.getTimestampRange());
                }

                @Override
                public void onFailure(Exception e) {
                    logger.debug("{} recovery failed after being marked as done", this);
                }
            }, this::decRef));
        }
    }
```

(we could call `onRecoveryFailure` on failure here too perhaps, although I think we don't today)

Reply:
I made this change and added a test for it.